src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs

```diff
@@ -306,7 +306,18 @@ public override TypeSyntax VisitPointerType(IPointerTypeSymbol symbol)
 
             public override TypeSyntax VisitTypeParameter(ITypeParameterSymbol symbol)
             {
-                return AddInformationTo(symbol.Name.ToIdentifierName(), symbol);
+                TypeSyntax typeSyntax =  AddInformationTo(symbol.Name.ToIdentifierName(), symbol);
+
+                // Add a ? if the type parameter is annotated as nullable, and the constraints allow it. It's possible to end up with
+                // a T? where T is constrained to class? if we combined a type parameter with some flow state -- we'll ignore the ? here
+                // at this stage. We should consider moving that check earlier when we actually wrap the symbol.
```

Comment:
I recognize this check isn't right -- I'm just getting things off the ground to see what else might be breaking here.

Reply:
Can you elaborate when this would be incorrect? Is it the case where flow state shows that we don't need to mark as nullable only because null will never be passed in?

Reply:
The check for "HasReferenceTypeConstraint" is probably insufficient, because you can be constrainted to reference types if you have something like "T : SomeClass". And there's probably other cases too.